Display errors inline rather than in toast

grzesiek2010 opened the pull request:
Closes #5618 
Closes #5751 

#### What has been done to verify that this works as intended?
I've tested the changes manually and fixed automated tests.

#### Why is this the best possible solution? Were any other approaches considered?
This is what we agreed on in the issue and when it comes to the implementation there is nothing to discuss here.

The final solution differs a little bit from what was described in the issue so let me list the changes that we introduced:

- Added a red outline with rounded corners to highlight the error. The outline should be displayed no matter what the question type is but there should be a small difference in displaying the error message when it's a text question. See the issue.
- Removed the red animation that we used to highlight required/invalid answers when there were multiple questions on the same page
- Updated the look of text fields. See: https://github.com/getodk/collect/pull/5725#issuecomment-1710785311 that's why testing read-only text widgets (plus external widgets and the Bearing widget too) is so important. See https://github.com/getodk/collect/issues/5751

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

Please test how different widgets look like when they are required but there is no answer or when constraints are violated. It's also important to keep in mind that apart from the default error messages we have when an error is displayed we can set custom ones too using the `required_message` and `constraint_message` columns in xls (see https://xlsform.org/en/#required-message and https://xlsform.org/en/#constraint-message)

#### Do we need any specific form for testing your changes? If so, please attach one.
I used the `All widgets` form but with all questions required to test every single widget:
[All widgets.xml.txt](https://github.com/getodk/collect/files/12607315/All.widgets.xml.txt)

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/getodk/docs/issues/new) and include the link below.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/docs/CODE-GUIDELINES.md#ui-components-style-guidelines)

grzesiek2010 commented:
@alyblenkin, it would be good to have early feedback from you. You can download the apk from CircleCI and here I'm attaching two forms I've used so far:
[all-widgets.xlsx](https://github.com/getodk/collect/files/12524188/all-widgets.xlsx) - all questions are required so you can see how it looks like with different widgets.
[RequiredAndConstraint.xlsx](https://github.com/getodk/collect/files/12524189/RequiredAndConstraint.xlsx) - it contains only 4 questions and two of them have custom messages (see https://xlsform.org/en/#required-message and https://xlsform.org/en/#constraint-message)

I've already noticed that some widgets will need to be updated because even if they do not have answers they still keep room for them visible for example:

![Screenshot_1693920480](https://github.com/getodk/collect/assets/3276264/6e579118-2c41-42db-876a-140332c0d2b0)

List widgets also would need to be updated because they don't have any margin end/right and some elements are visible out of the red outline:
 
![Screenshot_1693920509](https://github.com/getodk/collect/assets/3276264/44fd1a4b-22f2-4d1d-b8a2-2dcbce5b3461)

what do you think about those two? Any other thoughts?

alyblenkin commented:
Hey @grzesiek2010, I haven't been feeling well so I haven't gone through all of the widgets, but it's going in the right direction! My first reaction was we need a bit more padding between the question and the red line because it feels a bit squishy. However, we don't want the error to take up too much space, especially for smaller phones – so what you've done works.

I agree we would want to reduce the spacing on the ones that don't have answers. And fix the margin for the list widgets, so things don't overlap. For text field errors, I thought we should use https://m3.material.io/components/text-fields/specs#f157ea73-f2fd-4d99-aab3-c96bd9c91eec 

A few interaction things I noticed: 

The error doesn't go away once you've addressed the error in some cases, like when I uploaded the image.

![Screenshot_20230905-231445_ODK Collect](https://github.com/getodk/collect/assets/7871340/e8ef2a82-7afa-4b8c-9c50-d9529a8c0114)

I also noticed the error message disapears when you rotate the device. But it works if you have the device already rotated and try to go ahead without answering the question. 

I assume this is another issue, but the toast message is truncated here, which isn't ideal. 

![Screenshot_20230905-233208_ODK Collect](https://github.com/getodk/collect/assets/7871340/e0551464-df74-457b-970f-e9c8a7e6b7df)

grzesiek2010 commented:
> I assume this is another issue, but the toast message is truncated here, which isn't ideal.

I'm not sure if this is related to this issue that was supposed to be about required questions and constraints. That toast is displayed when a user tries to start an external app (using external widgets) or widgets like image/video but for some reason, there is no app installed on their device that can handle it. Probably it's something you can explore later and maybe we should get rid of those toasts too and display them in a similar way like constraints but it's a separate thing I would say.

grzesiek2010 committed:
Updated widgets to not display empty area for answers when there is no answer

grzesiek2010 commented:
>  For text field errors, I thought we should use https://m3.material.io/components/text-fields/specs#f157ea73-f2fd-4d99-aab3-c96bd9c91eec

Yeah I know but in the issue, we discussed implementing the same style for all widgets first and then (maybe in a separate pr) handling text widgets.

grzesiek2010 commented:
> My first reaction was we need a bit more padding between the question and the red line because it feels a bit squishy

I thought we wanted to avoid increasing the margins, and I haven't done that; I've simply introduced an outline between the screen edge and the questions. Do you think the outline should be shifted slightly towards the screen edge without expanding the margin between the edge and the questions? Or maybe you want to add more margin?

grzesiek2010 committed:
Display red outline to mark questions with errors

grzesiek2010 committed:
Do not display constraint toasts

grzesiek2010 committed:
Added the error layout to list widgets too

grzesiek2010 committed:
Moved the error layout to one shared file

grzesiek2010 committed:
Added margin to list widgets

grzesiek2010 committed:
Scroll to the bottom of the question with error

alyblenkin commented:
> > My first reaction was we need a bit more padding between the question and the red line because it feels a bit squishy
> 
> I thought we wanted to avoid increasing the margins, and I haven't done that; I've simply introduced an outline between the screen edge and the questions. Do you think the outline should be shifted slightly towards the screen edge without expanding the margin between the edge and the questions? Or maybe you want to add more margin?

Yes, we absolutely want to avoid increasing the margins. If it isn't too much work, I would be curious to see what it looks like slightly shifted towards the screen edge, but I also think what you have works!

alyblenkin commented:
> > I assume this is another issue, but the toast message is truncated here, which isn't ideal.
> 
> I'm not sure if this is related to this issue that was supposed to be about required questions and constraints. That toast is displayed when a user tries to start an external app (using external widgets) or widgets like image/video but for some reason, there is no app installed on their device that can handle it. Probably it's something you can explore later and maybe we should get rid of those toasts too and display them in a similar way like constraints but it's a separate thing I would say.

Yes, I think it's a different issue as well. I agree we should do away with the toast and make it more actionable as well, so the user knows what to do to resolve it. I'll add this to my backlog to investigate!

grzesiek2010 committed:
Fixed widget tests

grzesiek2010 committed:
Display errors withing textfields for text widgets

grzesiek2010 commented:
> The error doesn't go away once you've addressed the error in some cases, like when I uploaded the image.

Handling this properly (deciding if the error should be still visible or not)  would require form validation every time we receive a new answer. I'm not sure if it's a good idea because it would slow down the process. Maybe just removing the error in such cases would be enough (when we start an external app) then when a user is back the error would appear again on moving forward (if needed)?
But I'm not sure... I'm also ok with leaving this as-is. If you have other types of questions the error message won't disappear immediately when for example in select_one question you select something so why would we need to do that for widgets that start external apps?

grzesiek2010 commented:
> For text field errors, I thought we should use https://m3.material.io/components/text-fields/specs#f157ea73-f2fd-4d99-aab3-c96bd9c91eec

You can test it now @alyblenkin. There is only one thing... I used [outlined text fields](https://m3.material.io/components/text-fields/specs#f157ea73-f2fd-4d99-aab3-c96bd9c91eec) because I think they look better. You can install the app and let me know if you agree.

grzesiek2010 commented:
> I agree we would want to reduce the spacing on the ones that don't have answers. And fix the margin for the list widgets, so things don't overlap.

This is also ready so you can test it if you want.

alyblenkin commented:
> > The error doesn't go away once you've addressed the error in some cases, like when I uploaded the image.
> 
> Handling this properly (deciding if the error should be still visible or not) would require form validation every time we receive a new answer. I'm not sure if it's a good idea because it would slow down the process. Maybe just removing the error in such cases would be enough (when we start an external app) then when a user is back the error would appear again on moving forward (if needed)? But I'm not sure... I'm also ok with leaving this as-is. If you have other types of questions the error message won't disappear immediately when for example in select_one question you select something so why would we need to do that for widgets that start external apps?

I can see how that would slow down performance. I think removing the error would be helpful feedback for any type of question, so the user knows it's no longer invalid. I think this wasn't as obvious before because we just had the toast, now it's very clear something is wrong and the expected behaviour would be that it goes away once I've addressed the error.

alyblenkin commented:
> You can test it now @alyblenkin. There is only one thing... I used [outlined text fields](https://m3.material.io/components/text-fields/specs#f157ea73-f2fd-4d99-aab3-c96bd9c91eec) because I think they look better. You can install the app and let me know if you agree.

Testing this now :) I think the outline is growing on me and it's super clear what the user needs to engage with! We may want to add a bit more spacing between the question and the text field because it's a bit tight. 

![Screenshot_20230907-163529_ODK Collect](https://github.com/getodk/collect/assets/7871340/37ad669f-1378-484f-b7d1-d77ac87d91db)

I also noticed the outline text field is nearly touching the red error border here 
![Screenshot_20230907-163124_ODK Collect](https://github.com/getodk/collect/assets/7871340/b2391615-2ecb-47d9-8050-b2c89069bfe3)

I also don't think the list widget was updated because everything is still overlapping.

alyblenkin commented:
> > I agree we would want to reduce the spacing on the ones that don't have answers. And fix the margin for the list widgets, so things don't overlap.
> 
> This is also ready so you can test it if you want.

This looks much better!

alyblenkin commented:
@grzesiek2010 I was wondering if we could also try using the [surface container for the text field](https://m3.material.io/components/text-fields/specs#544d001a-e0ce-49eb-b77b-89fe08acccc5)? For the text field colour, we could use light grey. I think that might be a bit more subtle than the outline. 

This will also help us move towards our goal of creating consistency across the tools. Central uses this pattern already, and we would look to the same with Enketo. Central doesn't have a small radius on the corners, but we could move in that direction so everything is rounded. 

<img width="820" alt="Screenshot 2023-09-07 at 2 08 57 PM" src="https://github.com/getodk/collect/assets/7871340/2495e224-f6df-47bf-be53-a6fb9d9c9e2e">

grzesiek2010 commented:
> @grzesiek2010 I was wondering if we could also try using the [surface container for the text field](https://m3.material.io/components/text-fields/specs#544d001a-e0ce-49eb-b77b-89fe08acccc5)? For the text field colour, we could use light grey. I think that might be a bit more subtle than the outline.

Done. You can test it.

grzesiek2010 commented:
> I also don't think the list widget was updated because everything is still overlapping.

I only added the margin end so that the red outline doesn't overlap the labels. Those list widgets have labels and choices placed horizontally and are rather supposed to be used on larger screens otherwise the choices/labels will overlap.

grzesiek2010 commented:
> I can see how that would slow down performance. I think removing the error would be helpful feedback for any type of question, so the user knows it's no longer invalid. I think this wasn't as obvious before because we just had the toast, now it's very clear something is wrong and the expected behaviour would be that it goes away once I've addressed the error.

I've added removing the error every time the answer changes (no matter what widget it is) please play with it and let me know if you like it. The fact that there is a new answer doesn't mean there should be no error. We would need to validate constraints to check it but maybe it's better than assuming that the new answer is still wrong and removing the error only after navigating forward. So basically there is no perfect way but I think I like what I've implemented now a little bit more.

grzesiek2010 commented:
> I also noticed the error message disapears when you rotate the device

If you asked me to fix this a few years ago I would say of course but my thinking has changed a little bit over the past few years and now I'm more into keeping only crucial things like answers etc. after rotating a device. We have talked about that with @seadowg a few times and I think we were generally on the same page. 
@seadowg is this error something that you would rather keep or you agree it's not a must and maybe we can add it later if we change our mind?

alyblenkin commented:
> > @grzesiek2010 I was wondering if we could also try using the [surface container for the text field](https://m3.material.io/components/text-fields/specs#544d001a-e0ce-49eb-b77b-89fe08acccc5)? For the text field colour, we could use light grey. I think that might be a bit more subtle than the outline.
> 
> Done. You can test it.

Thanks so much for trying this out! I prefer this option over the outline. Could we make the grey a bit lighter? Perhaps the same colour as the main menu buttons? 

The spacing between the question and text field looks a lot better too!

alyblenkin commented:
> > I can see how that would slow down performance. I think removing the error would be helpful feedback for any type of question, so the user knows it's no longer invalid. I think this wasn't as obvious before because we just had the toast, now it's very clear something is wrong and the expected behaviour would be that it goes away once I've addressed the error.
> 
> I've added removing the error every time the answer changes (no matter what widget it is) please play with it and let me know if you like it. The fact that there is a new answer doesn't mean there should be no error. We would need to validate constraints to check it but maybe it's better than assuming that the new answer is still wrong and removing the error only after navigating forward. So basically there is no perfect way but I think I like what I've implemented now a little bit more.

I really like what you implemented. I think this is more consistent with what the user would expect. Before, I kept thinking my new answer was wrong because the error message persisted. I think it makes more sense to validate the new answer once they try to move forward. 

I also noticed you added the field text error message! It looks great :) 

Small note: It looks like the bearing widget is still using the underline, instead of the new text field with the surface colour in the background.

grzesiek2010 committed:
Hide error when answer is changed

grzesiek2010 committed:
Fixed widgets that didn't handle widgetValueChanged() properly

grzesiek2010 committed:
Changed TextInputLayout background color

grzesiek2010 commented:
> Thanks so much for trying this out! I prefer this option over the outline

Done.

> Could we make the grey a bit lighter? Perhaps the same colour as the main menu buttons?

Done.

> Small note: It looks like the bearing widget is still using the underline, instead of the new text field with the surface colour in the background.

Fixed.

@alyblenkin Do we need to change anything else?

alyblenkin commented:
> @alyblenkin Do we need to change anything else?

Not that I can think of! I think we are good to go :)

grzesiek2010 committed:
Fixed tests

grzesiek2010 commented:
Ok so two questions from me:

In the issue, you said:
> Currently, the error fades in when you have a group of questions, which we want to keep inside the bounds of the outline. This might be something we want to pair on to get right.

are you ok with the fact that I removed that old animation? I think the outline is enough.

What do you think about displaying answers in widgets that should start an external app or use a sensor but it's not available and then we want to allow adding answers manually? It's always been handled using text fields but those text fields were disabled by default and their background were removed so that initially they looked like text views. I think we wanted that because if the external app existed we wanted the answer to look like in any other widget with buttons + text (eg. geo widgets, barcode widget etc.). But if the external app is not available we wanted to allow adding the answer manually and enable that text field. In this pr I didn't implement that first part (removing the background and making it look like a text view).
![Screenshot_1694680607](https://github.com/getodk/collect/assets/3276264/1d7d9280-05e6-41fd-85c4-5b7ca4e5db26)
Maybe it's something for a separate issue if we want it? I'm kind of torn.

seadowg commented:
> If you asked me to fix this a few years ago I would say of course but my thinking has changed a little bit over the past few years and now I'm more into keeping only crucial things like answers etc. after rotating a device. We have talked about that with @seadowg a few times and I think we were generally on the same page.
@seadowg is this error something that you would rather keep or you agree it's not a must and maybe we can add it later if we change our mind?

Good question! Without looking at/thinking about the code, I think with this one I'd lean towards it not needing to if it's more work. We could always add it later, and my assumption is that rotating the device while fixing an error would be rare.

grzesiek2010 commented:
@seadowg 
I was also thinking about what I said in the issue regarding reworking that asterisk mark displayed when a question is required. Initially, I thought that adding a separate view for that would be better instead of treating it as a part of the text itself but it's not as easy and perfect as I thought. The problem would be that with long question labels, there would be a margin not only in the first line (where the view would be placed) but also in other lines. It would look like:
![Screenshot_1694643205](https://github.com/getodk/collect/assets/3276264/56566afd-9a00-438c-9e65-d8bd6aa34b87)
so I'm not going to do this here because we probably need to think of it more. If you have any thoughts (or @alyblenkin ) please share.

alyblenkin commented:
> are you ok with the fact that I removed that old animation? I think the outline is enough.

Yes, I think the outline is enough! We can always iterate on it later. 

> What do you think about displaying answers in widgets that should start an external app or use a sensor but it's not available and then we want to allow adding answers manually? It's always been handled using text fields but those text fields were disabled by default and their background were removed so that initially they looked like text views. I think we wanted that because if the external app existed we wanted the answer to look like in any other widget with buttons + text (eg. geo widgets, barcode widget etc.). But if the external app is not available we wanted to allow adding the answer manually and enable that text field. In this pr I didn't implement that first part (removing the background and making it look like a text view).

Yes, I think allowing people to answer manually is perfect so they can still achieve their task. We can always add the background to the text field in another issue. As long as they can still answer manually I think that's good for this issue. 

I noticed the text field cursor becomes white when there is an error, is it possible for it to be blue?

alyblenkin commented:
> @seadowg I was also thinking about what I said in the issue regarding reworking that asterisk mark displayed when a question is required. Initially, I thought that adding a separate view for that would be better instead of treating it as a part of the text itself but it's not as easy and perfect as I thought. The problem would be that with long question labels, there would be a margin not only in the first line (where the view would be placed) but also in other lines. It would look like: !

> so I'm not going to do this here because we probably need to think of it more. If you have any thoughts (or @alyblenkin ) please share.

This might be a silly idea and not ideal visually, but I guess the asterisk could go above the textbox for longer questions temporarily if it's tricky to implement right now. Perhaps that's not possible but I wanted to put the idea out there :)

alyblenkin commented:
Hi @grzesiek2010, following up on some of the feedback from the call earlier. 

**1. The supporting text for the error message is VERY small 🐜**
Could we make it the same size as the widget error messages, so it's consistent? 

**2. Extra spacing because we don't display the label in the text field**
I think we should evenly distribute the spacing within the text box, like the image below. If we feel that looks like it's still taking up too much space, we could reduce the height, but we don't want the touch target to be too small, so I think we should try this first.

<img width="458" alt="Screenshot 2023-09-19 at 8 56 33 AM" src="https://github.com/getodk/collect/assets/7871340/6d3ff410-9173-4b00-a8f5-b6773dc1c058">

grzesiek2010 committed:
Fixed error text size and layout padding in text widgets

grzesiek2010 committed:
Use the same TextInputLayout in BearingWidget and other text widgets

grzesiek2010 committed:
Naming improvements

grzesiek2010 committed:
Fixed BearingWidgetTest

grzesiek2010 commented:
> I noticed the text field cursor becomes white when there is an error, is it possible for it to be blue?

I haven't experienced anything like that. What android version are you using?

> 1. The supporting text for the error message is VERY small 🐜
Could we make it the same size as the widget error messages, so it's consistent?

Fixed.

> 2. Extra spacing because we don't display the label in the text field
I think we should evenly distribute the spacing within the text box, like the image below. If we feel that looks like it's still taking up too much space, we could reduce the height, but we don't want the touch target to be too small, so I think we should try this first.

Fixed.

alyblenkin commented:
The new fixes look great 🎉 

I'm using Android version 12 
![Screenshot_20230920-222437_ODK Collect 1](https://github.com/getodk/collect/assets/7871340/d9fe3b55-51f1-483b-ab46-184bd3f42614)

lognaturel commented:
It's looking great! With [this form](https://docs.google.com/spreadsheets/d/13pdMv3wNbd14hjN-dqxXgmgLN0fzvfGxbQa6OXvJNpU/edit#gid=0), when I swipe forward without filling in name, it generally scrolss "What is your birthdate" to the top. I see the same white cursor that @alyblenkin mentioned, also Android 12.

grzesiek2010 commented:
> It's looking great! With [this form](https://docs.google.com/spreadsheets/d/1tfrpIrixbOHdAb_aLejWq7HTeJaSjZVqcQLdAP86WA0/edit#gid=0), when I swipe forward without filling in name, it generally scrolss "What is your birthdate" to the top.

Are you sure you have attached the form you wanted? This one has no constraints and only one required question and I can't see any question with `What is your birthdate`.

grzesiek2010 committed:
Fixed cursor error color

grzesiek2010 committed:
Make sure text fields look like text views when are disabled

grzesiek2010 commented:
> I see the same white cursor that @alyblenkin mentioned, also Android 12.

Fixed.

> This might be a silly idea and not ideal visually, but I guess the asterisk could go above the textbox for longer questions temporarily if it's tricky to implement right now. Perhaps that's not possible but I wanted to put the idea out there :)

I'm not sure if this would look good. Let's leave it for now I thought it would be easier and more beneficial for us but it's not the case. Of course you can think about it and file a separate issue if you think there is anything we can improve there.

grzesiek2010 commented:
> Yes, I think allowing people to answer manually is perfect so they can still achieve their task. We can always add the background to the text field in another issue. As long as they can still answer manually I think that's good for this issue.

eventually I've fixed that because it was a problem not only with external widgets but with normal but read-only text widgets too.
@lognaturel thanks for asking abut that!
The only difference now is that if we have a read-only text widget the answer is displayed with more margin (because the material text filed has more padding):

| master  | this branch |
| ------------- | ------------- |
| ![Screenshot_1695297266](https://github.com/getodk/collect/assets/3276264/e7a25139-b085-47aa-bc05-28b243de4ea5) | ![Screenshot_1695298461](https://github.com/getodk/collect/assets/3276264/5ee3a4eb-fe4d-45d0-99d9-88703178e1d4)  |

@lognaturel do you think it's a problem? If so I would consider doing something else and displaying answers in text views (not text fields) if the text field is disabled so that we don't need all those customization we add to text fields. That's not perfect too because we would need to keep two elements (text field and text view) and display only one depending on the conditions but probably a little bit easier that forcing text fields to look like text views if they are disabled.

lognaturel commented:
> @lognaturel do you think it's a problem?

😬 It does feel distracting to me. Might be easier to handle as a follow-on issue? Either way is fine.

grzesiek2010 commented:
> It does feel distracting to me. Might be easier to handle as a follow-on issue? Either way is fine.

Yes we can have it in a separate pr. I can even start working on it on top of this pr and then rebase.

lognaturel commented:
> Are you sure you have attached the form you wanted? 

Sorry, [this form](https://docs.google.com/spreadsheets/d/13pdMv3wNbd14hjN-dqxXgmgLN0fzvfGxbQa6OXvJNpU/edit#gid=0)

grzesiek2010 committed:
Scroll to the top of a question with error if it exists

grzesiek2010 commented:
> It's looking great! With [this form](https://docs.google.com/spreadsheets/d/13pdMv3wNbd14hjN-dqxXgmgLN0fzvfGxbQa6OXvJNpU/edit#gid=0), when I swipe forward without filling in name, it generally scrolss "What is your birthdate" to the top.

> Sorry, [this form](https://docs.google.com/spreadsheets/d/13pdMv3wNbd14hjN-dqxXgmgLN0fzvfGxbQa6OXvJNpU/edit#gid=0)

Ah ok now it's fixed.

grzesiek2010 commented:
@alyblenkin 
there is one thing from your description that I have't implemented because it's not easy:

> If the question is really long, we want to scroll the user to the bottom of the page if there is an error (similar to what we do for multiple questions).

If you feel that it's something important please file a separate issue.

alyblenkin commented:
@grzesiek2010, the new changes look good - thanks for fixing the cursor. I'll chat with Hélène about scrolling to the end of long questions and see when/ if we still want to implement it. I don't think it's pressing right now.

grzesiek2010 committed:
Added basic custom view for displaying tet answers

grzesiek2010 committed:
Updated StringWidget to use WidgetAnswerText

grzesiek2010 committed:
Updated DecimalWidget to use WidgetAnswerText

grzesiek2010 committed:
Updated IntegerWidget to use WidgetAnswerText

grzesiek2010 committed:
Updated StringNumberWidget to use WidgetAnswerText

grzesiek2010 committed:
Utilize the already existing setAnswer method in WidgetAnswerText

grzesiek2010 committed:
Updated ExDecimalWidget to use WidgetAnswerText

grzesiek2010 committed:
Updated ExIntegerWidget to use WidgetAnswerText

grzesiek2010 committed:
Updated ExStringWidget to use WidgetAnswerText

grzesiek2010 committed:
Call widgetValueChanged() when value in text field is changed

grzesiek2010 committed:
Updated tests to use WidgetAnswerText

grzesiek2010 committed:
Moved rest of the setup from StringWidget to WidgetAnswerText

grzesiek2010 committed:
Formatted code

grzesiek2010 committed:
Updated BearingWidget to use WidgetAnswerText

grzesiek2010 committed:
Fixed BearingWidgetTest

grzesiek2010 committed:
Added a comment

grzesiek2010 committed:
Do not call widgetValueChanged twice when answer is cleared or set

seadowg commented on collect_app/src/main/java/org/odk/collect/android/activities/FormFillingActivity.java:
```diff
@@ -556,15 +556,7 @@ private void setupViewModels(FormEntryViewModelFactory formEntryViewModelFactory
-                    createErrorDialog(new FormError.NonFatal(e.getMessage()));
-                }
-
+                getCurrentViewIfODKView().setErrorForQuestionWithIndex(failedValidationResult.getIndex(), formEntryViewModel.getErrorMessage(failedValidationResult, this));
```
I'm generally really liking the structure of all this, but before I get into all the details: I was wondering if it would make sense to have the error message be part of the `FailedValidationResult` rather than having to reach back in the `FormEntryViewModel` to get it here?

grzesiek2010 replied:
Not really. The only place where we construct `FailedValidationResult` is `JavaRosaFormController` because that's where the validation takes place. On one hand, it would be good to move the method responsible for generating error messages there since first that method calls some javarosa methods to check if there is a custom error message but if not it needs to read our strings, and to do that we need `context` that would need to be passed through a few methods and that wouldn't be easy.

seadowg replied:
Ah good point. I still think it'd be better to have the observation not have to reach back into the `FormEntryViewModel` to simplify the amount of interactions here. Maybe instead of `getValidationResult` returning the JavaRosa type, it could return a `Result<Unit, String>?` so the `FormFillingActivity` side would look like:

```kotlin
formEntryViewModel.getValidationError().observe(this, consumable -> {
    if (consumable.isConsumed()) {
        return;
    }
    String result = consumable.getValue();
    if (result instance Result.Error) {
        getCurrentViewIfODKView().setErrorForQuestionWithIndex(failedValidationResult.getIndex(), ((Result.Error) result.value));
        swipeHandler.setBeenSwiped(false);
    } else if (result insanceof Result.Success) {
        SnackbarUtils.showLongSnackbar(findViewById(R.id.llParent), getString(org.odk.collect.strings.R.string.success_form_validation), findViewById(R.id.buttonholder));
    }

    consumable.consume();
});
```

That way the `FormEntryViewModel` API is smaller, and we don't expose JavaRosa types from it.

grzesiek2010 replied:
but I need `FormIndex` here. In your example, it's `failedValidationResult.getIndex()` but `failedValidationResult` does not exist anymore. I would rather need a wrapper around `ValidationResult` that would add the message and then I would be able to use it in a similar way as you described. Or maybe I'm missing something?

seadowg replied:
> I would rather need a wrapper around ValidationResult that would add the message and then I would be able to use it in a similar way as you described. Or maybe I'm missing something?

Ah sorry that was a mistake on my part. I like the idea of a wrapper that `FormEntryViewModel` owns though that gives you a message and an index.

grzesiek2010 replied:
Ah it would still require passing context through many different methods to be able to read strings. I gave it a try and I think it doesn't make sense. In the view model class I would need to edit and pass context to:
- `public void moveForward(HashMap<FormIndex, IAnswerData> answers)`
- `public void moveForward(HashMap<FormIndex, IAnswerData> answers, Boolean evaluateConstraints) `
- `public void moveBackward(HashMap<FormIndex, IAnswerData> answers)`
- `public boolean updateAnswersForScreen(HashMap<FormIndex, IAnswerData> answers, Boolean evaluateConstraints)`
- `public boolean saveScreenAnswersToFormController(HashMap<FormIndex, IAnswerData> answers, Boolean evaluateConstraints)`
- `public void validate()`

those methods are usually used in more than one place so that means a lot of changes. Do you agree that it doesn't make sense?

seadowg replied:
Yeah sorry I'd not fully considered the strings needing to be loaded. 

I guess the wrapper could include the data the UI needs to create the string. It looks to me like the error message is either a string we get from the `FormController` or an Android string so the wrapper could include either a `String` or an `int` string ID.

I think an even nicer structure would be to a have a sealed validation error type here where some include strings from JavaRosa:

```
sealed class ValidationError {
    data class Constraint(val message: String?) : ValidationError
    data class Required(val message: String?) : ValidationError
}
```

The UI could then show the message if it's non-null, or just load the string if it's null. This would keep string loading (and `Context`) out of the ViewModel and `FormController` interactions out of the UI.

grzesiek2010 replied:
Ok I've implemented something like that.

seadowg reviewed (changes requested):
See https://github.com/getodk/collect/pull/5725#discussion_r1347296755

grzesiek2010 committed:
Pass error message in FailedValidationResult

seadowg commented on collect_app/src/main/java/org/odk/collect/android/utilities/ThemeUtils.java:
```diff
@@ -122,6 +122,11 @@ public int getColorOnSurface() {
     }
 
+    @ColorInt
+    public int getColorSurface() {
```
I don't think this is being used?

grzesiek2010 replied:
Right, fixed.

seadowg commented on collect_app/src/main/res/layout/question_error_layout.xml:
```diff
@@ -0,0 +1,34 @@
+        android:layout_marginEnd="@dimen/margin_standard"
+        android:textColor="?colorError"
+        android:layout_gravity="center_vertical"
+        style="@style/TextAppearance.Collect.Subtitle1"
```
This should us a theme attribute rather than direct style reference. Also, `Subtitle1` is a Material 2 text appearance so we should probably use a Material 3 one instead here.

grzesiek2010 replied:
Fixed.

seadowg commented on collect_app/src/main/res/layout/widget_answer_text_field.xml:
```diff
@@ -0,0 +1,18 @@
+    android:paddingHorizontal="@dimen/margin_standard"
+    android:paddingTop="@dimen/margin_standard"
+    app:boxBackgroundColor="?colorSurfaceContainerLow"
+    app:errorTextAppearance="@style/TextAppearance.Collect.Subtitle1">
```
Again, this should use a theme attribute and a Material 3 text appearance.

grzesiek2010 replied:
Fixed.

seadowg commented on collect_app/src/main/java/org/odk/collect/android/widgets/BarcodeWidget.java:
```diff
@@ -73,13 +73,15 @@ protected View onCreateAnswerView(Context context, FormEntryPrompt prompt, int a
             binding.barcodeButton.setText(getContext().getString(org.odk.collect.strings.R.string.replace_barcode));
             binding.barcodeAnswerText.setText(answer);
         }
+        binding.barcodeAnswerText.setVisibility(binding.barcodeAnswerText.getText().toString().isBlank() ? GONE : VISIBLE);
```
It would be good to add tests for the widget behaviour updated in 7338895c391c7b0995f5b06ba254257a56fbf11f

grzesiek2010 replied:
Hmm, I'm torn... generally I agree but I think we need such tests for every single widget not only those that I have updated here. I think I will file a separate issue and maybe even try to fix it as a part of this pr but it shouldn't block reviewing and testing. Does that sound right?

seadowg replied:
Yeah I'm happy with that being a separate issue! Go ahead and file it and add it to the v2023.4 milestone.

grzesiek2010 replied:
The issue is ready https://github.com/getodk/collect/issues/5842

seadowg commented on collect_app/src/main/java/org/odk/collect/android/widgets/RatingWidget.java:
```diff
@@ -101,6 +101,7 @@ public IAnswerData getAnswer() {
     @Override
     public void clearAnswer() {
         binding.ratingBar1.setRating(0.0F);
+        widgetValueChanged();
```
Also good to add tests for the behaviour changes in a610987094d967a1b7348eca7540253164783ae8

grzesiek2010 replied:
Done.

seadowg reviewed (changes requested):
Looking good! Just a few tweaksI think it'd be nice to make, but this is ready for testing.

grzesiek2010 committed:
Removed unused methods from ThemeUtils

grzesiek2010 committed:
Added tests for WidgetValueChangedListener

grzesiek2010 committed:
Fixed error text appearance by using Material3 attribute

grzesiek2010 committed:
Display red outline in all types of questions

grzesiek2010 committed:
Fixed padding in text widgets so thet the errror message is not cut off

grzesiek2010 committed:
Hide the empty space when a widget is read-only and there is no answer

grzesiek2010 commented:
@seadowg I think it would be good to fix https://github.com/getodk/collect/issues/5751 in one pr so that we don't have to test similar things twice. There is a bunch of commits that address that issue and some that fix things you pointed out during your last review. I think the last commit you reviewed was https://github.com/getodk/collect/pull/5725/commits/10da040a2860bfdcf1c38bce4b7185436b5f24b3 so you can focus on newer ones.

grzesiek2010 committed:
Removed unused import

grzesiek2010 committed:
Fixed espresso tests

dbemke commented:
In audio widget (All widgets required form) the red error outline in shown after adding the audio.
![audioerror](https://github.com/getodk/collect/assets/56479916/786ffbe7-8de0-40e7-9dd1-cb9c97dd71b9)

dbemke commented:
In Select with images the red outline is "covered" by images on the right.
![selectImagesOutline](https://github.com/getodk/collect/assets/56479916/56e239c0-50dc-4070-a3be-1fa7a6bde608)

dbemke commented:
In selects the line under choices goes beyond the red outline.
![selectLines](https://github.com/getodk/collect/assets/56479916/6b0d1a14-b9f0-4177-bca2-2edb634f6fdd)

dbemke commented:
Is this the expected look of the field to enter e.g strings?
![field](https://github.com/getodk/collect/assets/56479916/9db45dde-6bc6-426f-a798-ffa0c6eaaa23)

srujner commented:
In Select Multiple widget with packed columns and no buttons, selected choice goes beyond the red outline.
![2](https://github.com/getodk/collect/assets/85225519/75068bc9-acf5-4f55-925e-bc4b9cfb8731)

grzesiek2010 committed:
Call widgetValueChanged when new audio is recorded

grzesiek2010 commented:
> In audio widget (All widgets required form) the red error outline in shown after adding the audio.

Fixed.

> Is this the expected look of the field to enter e.g strings?

Yes.

> In Select with images the red outline is "covered" by images on the right.

> In selects the line under choices goes beyond the red outline.

> In Select Multiple widget with packed columns and no buttons, selected choice goes beyond the red outline.

This will be fixed in https://github.com/getodk/collect/issues/5731 let's leave it for now as it is.

srujner commented:
**Tested with Success!**

Verified on device with Android 13

Verified cases:

- Forms with required questions;
- Forms with ReadOnly questions;
- Light and Dark mode;
- Minimize the app, rotate the screen;

dbemke commented:
**Tested with Success!**

Verified on device with Android 10
